From f4760db6d417c90c9a12eebab6b88048314d3b3b Mon Sep 17 00:00:00 2001
From: Aleksey Kasatkin <akasatkin@mirantis.com>
Date: Tue, 13 Sep 2016 13:08:22 +0300
Subject: [PATCH] Add fixed IPs to ns-dhcp

This adds fixed IPs corresponding to subnets without gateways to
ns-dhcp port on its creation. So that dnsmasq could get requests and
answer via that IP to instances when they look for DHCP server.

Change-Id: I4adbc1fa17b49d60b559966f7ccd29282352da3c
Closes-Bug: 1541490
---
 networking_calico/agent/dhcp_agent.py | 27 +++++++++++++++++++++++----
 1 file changed, 23 insertions(+), 4 deletions(-)

diff --git a/networking_calico/agent/dhcp_agent.py b/networking_calico/agent/dhcp_agent.py
index 19a7d3f..59b875a 100644
--- a/networking_calico/agent/dhcp_agent.py
+++ b/networking_calico/agent/dhcp_agent.py
@@ -74,11 +74,30 @@ class FakePlugin(object):
 
     """
 
+    def __init__(self, etcd):
+        self.etcd = etcd
+
     def create_dhcp_port(self, port):
         """Support the following DHCP DeviceManager calls.
 
         dhcp_port = self.plugin.create_dhcp_port({'port': port_dict})
         """
+        try:
+            LOG.debug("subnet_: read")
+            result = self.etcd.client.read(
+                SUBNET_DIR, recursive=True, timeout=5)
+            if result and getattr(result, "children", None):
+                nodes = result.children
+
+                for node in nodes:
+                    subnet_id = node.key.split("/")[-1]
+                    LOG.debug("subnet_: {0}".format(subnet_id))
+                    LOG.debug("subnet_: data: {0}".format(node))
+            else:
+                LOG.debug("subnet_: no result")
+        except etcd.EtcdKeyNotFound:
+            LOG.warning("subnet_: no data")
+
         LOG.debug("create_dhcp_port: %s", port)
         port['port']['id'] = 'dhcp'
 
@@ -513,14 +532,14 @@ class CalicoDhcpAgent(DhcpAgent):
         # DnsmasqRouted class.
         self.dhcp_driver_cls = DnsmasqRouted
 
+        # Watch etcd for any endpoint changes for this host.
+        self.etcd = CalicoEtcdWatcher(self)
+
         # Override the RPC plugin (i.e. proxy to the Neutron database)
         # with a fake plugin.  The DHCP driver code calls when it
         # wants to tell Neutron that it is creating, updating or
         # releasing the DHCP port.
-        self.plugin_rpc = FakePlugin()
-
-        # Watch etcd for any endpoint changes for this host.
-        self.etcd = CalicoEtcdWatcher(self)
+        self.plugin_rpc = FakePlugin(self.etcd)
 
     def run(self):
         """Run the EtcdWatcher loop."""
-- 
1.9.1

